network/p2p/network.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/metric"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+var (
+	_ validators.Connector = (*Network)(nil)
+	_ common.AppHandler    = (*Network)(nil)
+	_ NodeSampler          = (*Network)(nil)
+)
+
+// ClientOption configures Client
+type ClientOption interface {
+	apply(options *ClientOptions)
+}
+
+type clientOptionFunc func(options *ClientOptions)
+
+func (o clientOptionFunc) apply(options *ClientOptions) {
+	o(options)
+}
+
+// WithNodeSampler configures the sampling strategy for Client.AppRequestAny
+func WithNodeSampler(nodeSampler NodeSampler) ClientOption {
+	return clientOptionFunc(func(options *ClientOptions) {
+		options.NodeSampler = nodeSampler
+	})
+}
```

Comment:
How will this actually be used? isn't pretty much any node sampler going to want to be based on who we are currently connected to - which is caught by the `Network`?

Reply:
The plan is to use this with the `ValidatorSampler` to allow nodes to sample off the validator set instead of the connected peer set.

Reply:
I think it's reasonable to also consider adding a connected validator set as a field in `Network` in here so instead of using `WithNodeSampler(sampler NodeSampler)` a client can pass in a `WithValidatorSampling()` or something instead.